content/doc/developer/plugin-development/optional-dependencies.adoc

```diff
@@ -0,0 +1,165 @@
+---
+layout: developersection
+title: Optional dependencies
+---
+
+By default, dependencies are mandatory — if any of the dependencies are unavailable then the plugin will be disabled and can not be used.
+
+However, this is just the default. It is possible to declare dependencies to other plugins as optional. 
+In this way, even if some of the optional dependencies are unavailable, the plugin will continue to be loaded and executed.
+
+To declare an optional dependency, add the `<optional/>` tag to the declared dependency in the plugin `pom.xml`.
+
+[source,xml]
+----
+<dependencies>
+  <dependency>
+    <groupId>org.jvnet.hudson.plugins</groupId>
+    <artifactId>javanet-uploader</artifactId>
+    <version>1.5</version>
+    <optional>true</optional>
+  </dependency>
+</dependencies>
+----
+
+If any of the classes in the unavailable optional dependencies are used in the plugin a `NoClassDefError` will be thrown. 
+It is still your responsibility to cope with these errors (although Jenkins itself does some of this handling. 
+If your particular extension fails to load with `LinkageError` because of a missing dependency, it will gracefully disable just that extension and move on to try instantiating other extensions.
+
+You can mark your extension as an optional extension by changing the `@Extension` annotation to `@Extension(optional = true)` and Jenkins doesn't log any class loading errors when reading it.
+
+[source,java]
+----
+@Extension(optional = true)
+public static class YourDescriptor extends Descriptor<YourDescribable> {
+    @Override
+    public String getDisplayName() {
+        return "YourDisplayName";
+    }
+}
+----
+
+To determine if an optional dependency is installed, the Jenkins.getPlugin(String) method can be used.
+
+[source,java]
+----
+if (Jenkins.getInstance().getPlugin("javanet-uploader") != null) {
+    // use classes in the "javanet-uploader" plugin
+}
+----
+
+## Plugin statuses and behaviors
+
+|===
+| Status | getPlugin("dependee") != null | getPlugin("dependee") .getWrapper().isEnabled() | Load class from your plugin
+
+| Not installed | false | NPE | Fail
+| Installed without restart | true | true | Succeed
+| Installed without restart before your plugin | true | true | Succeed
+| Installed | true | true | Succeed
+| Uninstalled - not restarted | true | true | Succeed
+| Disabled - not restarted | true | false | Succeed
+| Disabled | false | NPE | Fail
+
+|===
+
+## Code causes class loading
+
+When your plugin optionally depends on other plugins, you have to access classes in those plugins only when you have checked those plugins are installed.
+
+In most cases, you can do that by testing plugin installations before executing codes that access those classes, as described in Dependencies among plugins#Optional dependencies.
```

Comment:
The `plugins#Optional` in this line will not resolve as a link and will not be understood by the reader.

```suggestion
In most cases, you can do that by testing the plugin is installed before executing code that accesses optionally loaded plugin classes.
```